connection.go

```diff
@@ -1437,39 +1436,145 @@ func (c *Conn) handleFrames(
 	}
 	handshakeWasComplete := c.handshakeComplete
 	var handleErr error
+	var frame wire.Frame
+	var err error
+	var l int
+	var frameType wire.FrameType
+
 	for len(data) > 0 {
-		l, frame, err := c.frameParser.ParseNext(data, encLevel, c.version)
+		frameType, l, err = c.frameParser.ParseType(data, encLevel)
 		if err != nil {
+			if err == io.EOF {
+				break
+			}
 			return false, false, nil, err
 		}
 		data = data[l:]
-		if frame == nil {
-			break
-		}
-		if ackhandler.IsFrameAckEliciting(frame) {
+
+		if ackhandler.IsFrameTypeAckEliciting(frameType) {
 			isAckEliciting = true
 		}
-		if !wire.IsProbingFrame(frame) {
+		if !wire.IsProbingFrameType(frameType) {
 			isNonProbing = true
 		}
-		if log != nil {
-			frames = append(frames, toLoggingFrame(frame))
-		}
-		// An error occurred handling a previous frame.
-		// Don't handle the current frame.
-		if handleErr != nil {
+
+		// Fast path: We're inlining common cases, to avoid using interfaces
+		if frameType.IsStreamFrameType() {
+			streamFrame, l, err := wire.ParseStreamFrame(data, frameType, c.version)
+			if err != nil {
+				return false, false, nil, err
+			}
+
+			data = data[l:]
+			if streamFrame == nil {
+				break
+			}
+
+			if log != nil {
+				frames = append(frames, toLoggingFrame(streamFrame))
+			}
+			// An error occurred handling a previous frame.
+			// Don't handle the current frame.
+			if handleErr != nil {
+				continue
+			}
+			err = c.streamsMap.HandleStreamFrame(streamFrame, rcvTime)
+			if err != nil {
+				if log == nil {
+					return false, false, nil, err
+				}
+				// If we're logging, we need to keep parsing (but not handling) all frames.
+				handleErr = err
+			}
 			continue
 		}
-		pc, err := c.handleFrame(frame, encLevel, destConnID, rcvTime)
-		if err != nil {
-			if log == nil {
+		//nolint:exhaustive // StreamFrame has already been handled. We now need to handle a few more hot paths,
+		// the rest will be handled by the default clause.
+		switch frameType {
+		case wire.AckFrameType, wire.AckECNFrameType:
+			ackFrame, l, err := c.frameParser.ParseAckFrame(frameType, data, encLevel, c.version)
+			// Fast path: We inline the frame handling logic, to avoid using interfaces
+			if err != nil {
 				return false, false, nil, err
 			}
-			// If we're logging, we need to keep parsing (but not handling) all frames.
-			handleErr = err
-		}
-		if pc != nil {
-			pathChallenge = pc
+			data = data[l:]
+			if log != nil {
+				frames = append(frames, toLoggingFrame(ackFrame))
+			}
+			// An error occurred handling a previous frame.
+			// Don't handle the current frame.
+			if handleErr != nil {
+				continue
+			}
+			err = c.handleAckFrame(ackFrame, encLevel, rcvTime)
+			if err != nil {
+				if log == nil {
+					return false, false, nil, err
+				}
+				// If we're logging, we need to keep parsing (but not handling) all frames.
+				handleErr = err
+			}
+			continue
+		case wire.DatagramNoLengthFrameType, wire.DatagramWithLengthFrameType:
```

Comment:
It's probably nicer to introduce a `wire.IsDatagramFrame`, similar to the `wire.IsStreamFrame` you defined.